text/layer-physical-formats.md

```diff
@@ -0,0 +1,87 @@
+2. What units of underlying data were changed from the underlying layer *in previous, uncompacted, segments of this
+   layer*, and the LSN at which that data last changed (for O(1) lookup into the next, more granular, layer).
+3. (Implied by the above: If the page we’re looking for is not in (1) or (2), we can search in the next lower layer)
+
+... where a ‘unit of underlying data’ is a postgresql block for the lowest (oldest / most precise / smallest keyspace)
+LSM segments, and the segment of the lower layer for the higher (wider) layers.
+
+### On fast localization of data #2
+
+When we *do* eventually find some of the data we need to restore the requested page version, then that data should be
+complete enough to restore the full page; i.e. for some WAL / delta records found in one file, we should not have to
+go to another file to request the base page version. This implies that each page with deltas in a delta file will also
+have a base image in that same delta file. *This indeed wastes some space (with a bad pathological ~ 400x worst case of
+a single minimal WAL record on that page in this layer), but this is important in the long run; as it saves us from
+cross-file dependencies.*
+
+### On disk space savings methods: Physical representation of page version metadata
+
+As Branches are always replayed in ascending order of LSN, we should store the WAL in ascending order as well; so that
+any retrieval of the records (once located) is sequential.
+
+To support fast `getpage@LSN`-requests; we can store the LSN seperated from the records they describe; so that this
+data can be compressed and accessed seperately, without a need for further seeking through the file for the next 
+record to find out you don't need that record because the LSN is out of the applicable range.
+
+When storing LSNs seperate from the records they describe, we can use common compression operations to reduce the
+size of this metadata.
+
+Wal record lengths are generally small, so we can use variable-length integer encoding to store this field. Depending
+on the encoding used we can save a few bytes per record, up to 12.5% of the overhead in the case of a single XlogRecord
+without further data.
+
+### On PageServer optimizations: only retain replayable records for some amount of time
+
+At some point, storing more deltas is not worth the effort; it becomes clear that we’re not going to see requests from
+Compute with LSN < X. At that point, we could replay all completed segments <X onto existing image layers; so that we
+can drop the segments locally and save local storage space.
+
+### On Pageserver optimizations: write amplification / image layers
+
+When we have no need for intermediate image layers between layers which include
+deltas (due to e.g. low overhead of finding the right page version), we can
+decrease write amplification of creating and maintaining image layers by
+maintaining one "base" image layer, whose pages are modified and updated when
```

Comment:
> whose pages are modified and updated

From my (very limited) S3 remote storage standpoint, I'd love to see what changes in our checkpointing model and how that affects the fact that now no already created files are ever updated.

For now, before LSM trees (or whatever else) is implemented, we're bundling new files from checkpoints together into a streaming (zstd) archive and rely on this "never update" file approach when uncompressing, skipping the files existing locally (except the timeline metadata file).
Due to the same assumptions, we compress the checkpoint files without checking either, since they can always increment.

So far, only this particular place seems worrying to me, but do you have any concerns?

Reply:
The main difference between the current (potential future) checkpoint code would be that
 1.) top-down checkpoints (initiated by influx of WAL) won't need to generate Image layers, and 
 2.) GC checkpoints would potentially update the bottom-most Image layer, then upload that layer with the new LSN attached. 

Note that the upload of the Image layer would not strictly be required as long as the intermediate layers still exist in S3; as we'd basically be 'inlining' the `base + gc_able_differences` in the `base_image_layer + gc_able_page_deltas + new_page_deltas = requested page version` formula.

Reply:
Right, so more logic, connected to GC.
We have nothing on that front in pageserver s3 sync now, but with the potential _updates_ we should adapt the s3 sync logic anyway, so I would love to see a "new GC and pageserver's local file implications" section if/when any further development happens, so that I could pick up those updates on my side.

(AFAIK, LSM approach is also updating some files on rebalancing? If so, we're on the right way with adjusting the existing code to such events)